src/catalog/tests/open.rs

```diff
@@ -764,3 +766,78 @@ async fn test_tombstone(
 
     assert_eq!(openable_state2.get_tombstone().await.unwrap(), Some(true));
 }
+
+#[mz_ore::test(tokio::test)]
+#[cfg_attr(miri, ignore)] //  unsupported operation: can't call foreign function `TLS_client_method` on OS `linux`
+async fn test_persist_fencing() {
+    async fn testcase(catalog: &str, reader: &str, expected: Result<(), ()>) {
+        let catalog_version = semver::Version::parse(catalog).unwrap();
+        let reader_version = semver::Version::parse(reader).unwrap();
+        let organization_id = Uuid::new_v4();
+        let mut persist_cache = PersistClientCache::new_no_metrics();
+
+        persist_cache.cfg.build_version = catalog_version.clone();
+        let persist_client = persist_cache
+            .open(PersistLocation::new_in_mem())
+            .await
+            .expect("in-mem location is valid");
+        let persist_openable_state = test_persist_backed_catalog_state_with_version(
+            persist_client.clone(),
+            organization_id.clone(),
+            catalog_version,
+        )
+        .await
+        .unwrap();
+        let _persist_state = Box::new(persist_openable_state)
+            .open(NOW_ZERO(), &test_bootstrap_args(), None, None)
+            .await
+            .unwrap();
+
+        persist_cache.cfg.build_version = reader_version.clone();
+        let persist_client = persist_cache
+            .open(PersistLocation::new_in_mem())
+            .await
+            .expect("in-mem location is valid");
+        let persist_openable_state = test_persist_backed_catalog_state_with_version(
+            persist_client.clone(),
+            organization_id.clone(),
+            reader_version,
+        )
+        .await
+        .map(|_| ())
+        .map_err(|err| match err {
+            DurableCatalogError::IncompatiblePersistVersion { .. } => (),
+            err => panic!("unexpected error: {err:?}"),
+        });
+        assert_eq!(
+            persist_openable_state, expected,
+            "test case failed, catalog: {catalog}, reader: {reader}, expected: {expected:?}"
+        );
+    }
+
+    // Test cases copied from mz_persist_client::internal::encoding::tests::check_data_versions.
```

Comment:
Duplicating the full set of tests cases feels like it will inevitably rot. Given that most of the logic is reused by sharing the `is_data_version_invalid` method between the two, I think this test would probably be clearer if we just picked a small number of interesting examples (to ensure that the plumbing to working as expected) instead of copying the full set. Maybe code=`0.10.0` and data=`0.{10,11,12}.0`?

Reply:
Good call, updated.